test/unittests/t_catalog_mgr_rw.cc

```diff
@@ -118,4 +131,98 @@ TEST_F(T_CatalogMgrRw, CloneTree) {
   EXPECT_TRUE(dirent.IsNestedCatalogRoot());
 }
 
+
+TEST_F(T_CatalogMgrRw, SwapNestedCatalog) {
+  CatalogTestTool tester("swap_nested_catalog");
+  EXPECT_TRUE(tester.Init());
+
+  DirSpec spec = MakeBaseSpec();
+  EXPECT_TRUE(tester.ApplyAtRootHash(tester.manifest()->catalog_hash(), spec));
+
+  catalog::WritableCatalogManager *catalog_mgr = tester.catalog_mgr();
+
+  // Look up sub1 and sub2 nested catalogs
+  PathString path;
+  shash::Any sub1_hash;
+  shash::Any sub2_hash;
+  uint64_t sub1_size;
+  uint64_t sub2_size;
+  EXPECT_TRUE(catalog_mgr->LookupNested(PathString("/dir/dir/dir/sub1"),
+                                        &path, &sub1_hash, &sub1_size));
+  EXPECT_TRUE(catalog_mgr->LookupNested(PathString("/dir/dir/dir/sub2"),
+                                        &path, &sub2_hash, &sub2_size));
+
+  // Swap sub1 with itself
+  DirectoryEntry dirent;
+  catalog_mgr->DetachNested();
+  catalog_mgr->SwapNestedCatalog("dir/dir/dir/sub1", sub1_hash, sub1_size);
+  EXPECT_TRUE(catalog_mgr->LookupPath("/dir/dir/dir/sub1/file1",
+                                      kLookupSole, &dirent));
+  EXPECT_STREQ(g_hashes[6], dirent.checksum().ToString().c_str());
+
+  // Swap sub1 and sub2
+  catalog_mgr->DetachNested();
+  catalog_mgr->SwapNestedCatalog("dir/dir/dir/sub1", sub2_hash, sub2_size);
+  catalog_mgr->SwapNestedCatalog("dir/dir/dir/sub2", sub1_hash, sub1_size);
+  EXPECT_TRUE(catalog_mgr->LookupPath("/dir/dir/dir/sub1/file2",
+                                      kLookupSole, &dirent));
+  EXPECT_STREQ(g_hashes[7], dirent.checksum().ToString().c_str());
+  EXPECT_TRUE(catalog_mgr->LookupPath("/dir/dir/dir/sub2/file1",
+                                      kLookupSole, &dirent));
+  EXPECT_STREQ(g_hashes[6], dirent.checksum().ToString().c_str());
+}
+
+
+TEST_F(T_CatalogMgrRw, SwapNestedCatalogFailSlow) {
+  CatalogTestTool tester("swap_nested_catalog_fail_slow");
+  EXPECT_TRUE(tester.Init());
+
+  DirSpec spec = MakeBaseSpec();
+  EXPECT_TRUE(tester.ApplyAtRootHash(tester.manifest()->catalog_hash(), spec));
+
+  catalog::WritableCatalogManager *catalog_mgr = tester.catalog_mgr();
+
+  // Look up sub1 and sub2 nested catalogs
+  PathString path;
+  shash::Any sub1_hash;
+  shash::Any sub2_hash;
+  uint64_t sub1_size;
+  uint64_t sub2_size;
+  EXPECT_TRUE(catalog_mgr->LookupNested(PathString("/dir/dir/dir/sub1"),
+                                        &path, &sub1_hash, &sub1_size));
+  EXPECT_TRUE(catalog_mgr->LookupNested(PathString("/dir/dir/dir/sub2"),
+                                        &path, &sub2_hash, &sub2_size));
+
+  // Create nonexistent nested catalog hash
+  shash::Any subX_hash(shash::kMd5,
+                       shash::HexPtr("3e25960a79dbc69b674cd4ec67a72c62"),
+                       shash::kSuffixCatalog);
+  uint64_t subX_size = 42;
+
+  // Fail if parent catalog does not exist
+  catalog_mgr->DetachNested();
+  EXPECT_DEATH(catalog_mgr->SwapNestedCatalog("/dir/dir/dir/sub1",
```

Comment:
It's currently subtle: the nested catalog does not exist due to the leading slash, right?

```suggestion
  EXPECT_DEATH(catalog_mgr->SwapNestedCatalog("/no/such/directory/sub1",
```

Reply:
You're right: I think this is a relic from when I was still being surprised by the unexpected behaviours of `MakeRelative()`/`MakeRelativePath()` (see https://github.com/cvmfs/cvmfs/pull/2709#issuecomment-853102465).

Will update.